src/Symfony/Bundle/SecurityBundle/Debug/DebugRoleHierarchy.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Debug;
+
+use Symfony\Component\Security\Core\Role\RoleHierarchyInterface;
+
+/**
+ * Wraps a RoleHierarchy to access computed map, hierarchy tree and placeholders.
+ *
+ * @author Nicolas Rigaud <squrious@protonmail.com>
+ *
+ * @internal
+ */
+final class DebugRoleHierarchy
+{
+    private array $map = [];
+    private array $hierarchy = [];
+    private array $placeholders = [];
+    private ?\Closure $placeholderMatcher = null;
+
+    public function __construct(private readonly RoleHierarchyInterface $decorated)
+    {
+        if (
+            property_exists($this->decorated, 'map')
+            && property_exists($this->decorated, 'hierarchy')
+            && property_exists($this->decorated, 'rolePlaceholdersPatterns')
+            && method_exists($this->decorated, 'getMatchingPlaceholders')
+        ) {
+            $this->map = (new \ReflectionProperty($this->decorated, 'map'))->getValue($this->decorated);
+            $this->hierarchy = (new \ReflectionProperty($this->decorated, 'hierarchy'))->getValue($this->decorated);
```

Comment:
Using Reflection to access private properties is a bad idea. Using reflection to access private properties in a different package is a nightmare, as there is strictly no BC guarantee for those (the fact that the property exist does not even mean that it has the type that you expect here, as it might be a custom implementation as well)

Reply:
I removed reflection stuff in favor of inheritance and a dedicated service for the debug purposes.